contracts/proxy-lib/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+
+// An internal request wrapped with the signer_pk and added timestamp to determine num_requests_pk and prevent against malicious key holder gas attacks
+#[derive(Clone, PartialEq)]
+#[near(serializers = [json, borsh])]
+pub struct ConfirmationRequestWithSigner {
+    pub proposal_id: ProposalId,
+    pub signer_id: Repr<SignerId>,
+    pub added_timestamp: u64,
+}
+
+/// Lowest level action that can be performed by the multisig contract.
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub enum ProposalAction {
+    FunctionCall {
+        method_name: String,
+        args: Base64VecU8,
+        deposit: U128,
+        gas: U64,
+    },
+}
+
+// The request the user makes specifying the receiving account and actions they want to execute (1 tx)
+#[derive(Clone, PartialEq, Debug)]
+#[near(serializers = [json, borsh])]
+pub struct Proposal {
+    pub receiver_id: AccountId,
+    pub author_id: Repr<SignerId>,
+    pub actions: Vec<ProposalAction>,
+}
+
+#[near]
+impl ProxyContract {
+    #[init]
+    pub fn init(context_id: Repr<ContextId>, context_config_account_id: AccountId) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            context_id,
+            context_config_account_id,
+            proposal_nonce: 0,
+            proposals: IterableMap::new(b"r".to_vec()),
+            approvals: IterableMap::new(b"c".to_vec()),
+            num_proposals_pk: IterableMap::new(b"k".to_vec()),
+            num_approvals: 2,
+            active_proposals_limit: 10,
+        }
+    }
+
+    pub fn create_and_approve_proposal(&self, proposal: Signed<Proposal>) -> Promise {
+        // Verify the signature corresponds to the signer_id
+        let proposal = proposal.parse(|i| *i.author_id).expect("failed to parse input");
+
+        let author_id = proposal.author_id.rt().expect("Invalid signer");
+
+        let num_proposals = self.num_proposals_pk.get(&author_id).unwrap_or(&0) + 1;
+        assert!(
+            num_proposals <= self.active_proposals_limit,
+            "Account has too many active proposals. Confirm or delete some."
+        );
+        return self.create_by_member(proposal, num_proposals)
+    }
+
+    pub fn approve(&mut self, request: Signed<ConfirmationRequestWithSigner>) -> Promise {
+        let request = request
+            .parse(|i| *i.signer_id)
+            .expect("failed to parse input");
+        return self.approve_by_member(request.signer_id, request.proposal_id);
+    }
+
+    fn internal_confirm(&mut self, request_id: ProposalId, signer_id: Repr<SignerId>) -> () {
+        let confirmations = self.approvals.get_mut(&request_id).unwrap();
+        assert!(
+            !confirmations.contains(&signer_id),
+            "Already confirmed this request with this key"
+        );
+        confirmations.insert(signer_id);
+
+        // Check if the number of confirmations is enough to execute the request
```

Comment:
As discussed, you'll be introducing a mechanism for re-validating the approvals, yeah?

So you'll need to keep a list of Signer IDs, and `has_member` will need to become `has_members`, yeah?

Or maybe pass a list of IDs, and return a list of non-members. So we can identify which approvals to invalidate.

Or instead of returning a list of non-members, it just returned a count of non-members. So knowing when to execute the request would be:

`if threshold == (number_of_members_checked - number_of_non_members_received) {}`

Reply:
discussed internally. This will be part of a new PR